torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -0,0 +1,111 @@
+import math
+
+import torch
+import torch.distributed as dist
+
+
+def _orthogonalize(matrix, epsilon=1e-8):
```

Comment:
Any reason we're not using a built-in orthogonalization algorithm such as one provided by numpy that is battle-tested?

Reply:
Let's use the paper author's implementation first, since I am not sure if there is any subtle difference between the two versions.

Reply:
You could use `torch.qr()` and keep the `Q` part. That should do the same thing (apart from the epsilon parameter, maybe a good unit test) 
When I first wrote this, the hand-crafted procedure was faster. Maybe the QR algorithm is not optimized for long and narrow matrices with very few columns

Reply:
> You could use `torch.qr()` and keep the `Q` part. That should do the same thing (apart from the epsilon parameter, maybe a good unit test)

In this implementation, epsilon is 0 when orthogonalizing both P and Q. Why do you suggest keeping the Q part?

> When I first wrote this, the hand-crafted procedure was faster. Maybe the QR algorithm is not optimized for long and narrow matrices with very few columns

Interesting! Since we expect the rank here is a very small number, typically 1 or 2, then probably it's better to go with your hand-crafted version for now.

Reply:
> Why do you suggest keeping the Q part?
Ah, with Q I meant the Q from Q-R decomposition, not the Q from the P and Q in PowerSGD.

Reply:
> > Why do you suggest keeping the Q part?
> 
> Ah, with Q I meant the Q from Q-R decomposition, not the Q from the P and Q in PowerSGD.

Thanks! I think you meant `matrix[:], _ = torch.qr(matrix)`